### DISCUSSION TITLE
Remove `correctness/noDelete` from the default/recommended rules

### DISCUSSION THREAD

**Discussion by lgarron** [rome/tools#3614] [Suggestions]
### Environment information

<details><pre><code>
block
CLI:
  Version:              10.0.0
  Color support:        true

Platform:
  CPU Architecture:     aarch64
  OS:                   macos

Environment:
  ROME_LOG_DIR:         unset
  NO_COLOR:             unset
  TERM:                 "xterm-256color"

Rome Configuration:
  Status:               loaded
  Formatter disabled:   false
  Linter disabled:      false

Workspace:
  Open Documents:       0

Discovering running Rome servers...

Running Rome Server: ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

i The client isn't connected to any server but rage discovered this running Rome server.

× Workspace rage failed: Method not found

Rome Server Log:

! Please review the content of the log file before sharing it publicly as it may contain sensitive information:
  * Path names that may reveal your name, a project name, or the name of your employer.
  * Source code

├─69202401ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─69202409ms INFO rome_lsp::server Starting Rome Language Server...
├─69202411ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─69202411ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/mastodon/mastodon/rome.json"
├─69202411ms ERROR rome_service::configuration Could not find the file configuration at "/Users/lgarron/Code/git/github.com/mastodon/mastodon/rome.json"
├─69202411ms ERROR rome_service::configuration Reason: Os { code: 2, kind: NotFound, message: "No such file or directory" }
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(2)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─┐rome_formatter::printer::Printer::print{}
├─┘
├─70450748ms WARN tower_lsp Got a textDocument/didSave notification, but it is not implemented
├─┐rome_js_parser::parse::parse{file_id=FileId(2)}
├─┘
├─70453443ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70453443ms INFO tower_lsp::service::layers exit notification received, stopping
├─70453485ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70453488ms INFO rome_lsp::server Starting Rome Language Server...
├─70453489ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70453489ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/cubing/cubing.js/rome.json"
├─┐rome_fs::fs::os::OsFile::read_to_string{}
├─┘
├─70453489ms INFO rome_lsp::session Configuration found, and it is valid!
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─70467808ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70467820ms INFO tower_lsp::service::layers exit notification received, stopping
├─70469518ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70469521ms INFO rome_lsp::server Starting Rome Language Server...
├─70469522ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70469522ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/cubing/cubing.js/rome.json"
├─┐rome_fs::fs::os::OsFile::read_to_string{}
├─┘
├─70469522ms INFO rome_lsp::session Configuration found, and it is valid!
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─┐rome_formatter::printer::Printer::print{}
├─┘
├─70610318ms WARN tower_lsp Got a textDocument/didSave notification, but it is not implemented
├─70693757ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70693771ms INFO tower_lsp::service::layers exit notification received, stopping
├─70693831ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70694546ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70694556ms INFO tower_lsp::service::layers exit notification received, stopping
├─70694889ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70694901ms INFO tower_lsp::service::layers exit notification received, stopping
├─70695585ms INFO tower_lsp::service::layers shutdown request received, shutting down
├─70695593ms INFO tower_lsp::service::layers exit notification received, stopping
├─70695602ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70700672ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70700682ms INFO rome_lsp::server Starting Rome Language Server...
├─70700718ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70700718ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/lgarron/trusted-types-playground/rome.json"
├─70700718ms ERROR rome_service::configuration Could not find the file configuration at "/Users/lgarron/Code/git/github.com/lgarron/trusted-types-playground/rome.json"
├─70700718ms ERROR rome_service::configuration Reason: Os { code: 2, kind: NotFound, message: "No such file or directory" }
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─70701052ms WARN tower_lsp::service::pending client asked to cancel request 1, but no such pending request exists, ignoring
├─70701724ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70701776ms INFO rome_lsp::server Starting Rome Language Server...
├─70701915ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70701915ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/cubing/timer/rome.json"
├─┐rome_fs::fs::os::OsFile::read_to_string{}
├─┘
├─70701916ms INFO rome_lsp::session Configuration found, and it is valid!
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─70704351ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70704372ms INFO rome_lsp::server Starting Rome Language Server...
├─70704394ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70704394ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/mastodon/mastodon/rome.json"
├─70704394ms ERROR rome_service::configuration Could not find the file configuration at "/Users/lgarron/Code/git/github.com/mastodon/mastodon/rome.json"
├─70704394ms ERROR rome_service::configuration Reason: Os { code: 2, kind: NotFound, message: "No such file or directory" }
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─70705148ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70705153ms INFO rome_lsp::server Starting Rome Language Server...
├─70705159ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70705160ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/cubing/alg.cubing.net/rome.json"
├─70705160ms ERROR rome_service::configuration Could not find the file configuration at "/Users/lgarron/Code/git/github.com/cubing/alg.cubing.net/rome.json"
├─70705160ms ERROR rome_service::configuration Reason: Os { code: 2, kind: NotFound, message: "No such file or directory" }
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─70743407ms WARN tower_lsp::service::pending client asked to cancel request 1, but no such pending request exists, ignoring
├─70743437ms WARN tower_lsp::service::pending client asked to cancel request 2, but no such pending request exists, ignoring
├─70743455ms INFO rome_lsp::server Starting Rome Language Server...
├─70743455ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70743468ms INFO rome_lsp::server Attempting to load the configuration from 'rome.json' file
├─70743468ms INFO rome_service::configuration Attempting to load the configuration file at path "/Users/lgarron/Code/git/github.com/cubing/cubing.js/rome.json"
├─┐rome_fs::fs::os::OsFile::read_to_string{}
├─┘
├─70743469ms INFO rome_lsp::session Configuration found, and it is valid!
├─┐rome_js_parser::parse::parse{file_id=FileId(0)}
├─┘
├─┐rome_js_parser::parse::parse{file_id=FileId(1)}
├─┘
├─70911321ms INFO rome_lsp::server Starting Rome Language Server...
├─70911322ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
├─70919313ms INFO rome_lsp::server Starting Rome Language Server...
├─70919314ms ERROR tower_lsp::transport failed to encode message: failed to encode response: Socket is not connected (os error 57)
</code></pre></details>

### What happened?

Rome errors with `This is an unexpected use of the delete operator.` for the following code:

```ts
interface AccountSettings {
  userID: string;
  displayName?: string;
}

const settings: AccountSettings = {
  userID: "lgarron",
  displayName: "Lucas Garron",
};
delete settings["displayName"]; // ←←←← this line!
if ("displayName" in settings) {
  console.log(`You have a display name, ${settings["displayName"]}!`);
} else {
  console.log("You don't have a display name, would you like to set one?");
}
```
[Playground link](https://play.rome.tools/?code=interface+AccountSettings+%7B%0A++userID%3A+string%3B%0A++displayName%3F%3A+string%3B%0A%7D%0A%0Aconst+settings%3A+AccountSettings+%3D+%7B%0A++userID%3A+%22lgarron%22%2C%0A++displayName%3A+%22Lucas+Garron%22%2C%0A%7D%3B%0Adelete+settings%5B%22displayName%22%5D%3B%0Aif+%28%22displayName%22+in+settings%29+%7B%0A++console.log%28%60You+have+a+display+name%2C+%24%7Bsettings%5B%22displayName%22%5D%7D%21%60%29%3B%0A%7D+else+%7B%0A++console.log%28%22You+don%27t+have+a+display+name%2C+would+you+like+to+set+one%3F%22%29%3B%0A%7D%0A&lineWidth=80&indentStyle=tab&quoteStyle=double&quoteProperties=as-needed&trailingComma=all&indentWidth=2&sourceType=module&enabledNurseryRules=true&typescript=true&jsx=false#aQBuAHQAZQByAGYAYQBjAGUAIABBAGMAYwBvAHUAbgB0AFMAZQB0AHQAaQBuAGcAcwAgAHsACgAgACAAdQBzAGUAcgBJAEQAOgAgAHMAdAByAGkAbgBnADsACgAgACAAZABpAHMAcABsAGEAeQBOAGEAbQBlAD8AOgAgAHMAdAByAGkAbgBnADsACgB9AAoACgBjAG8AbgBzAHQAIABzAGUAdAB0AGkAbgBnAHMAOgAgAEEAYwBjAG8AdQBuAHQAUwBlAHQAdABpAG4AZwBzACAAPQAgAHsACgAgACAAdQBzAGUAcgBJAEQAOgAgACIAbABnAGEAcgByAG8AbgAiACwACgAgACAAZABpAHMAcABsAGEAeQBOAGEAbQBlADoAIAAiAEwAdQBjAGEAcwAgAEcAYQByAHIAbwBuACIALAAKAH0AOwAKAGQAZQBsAGUAdABlACAAcwBlAHQAdABpAG4AZwBzAFsAIgBkAGkAcwBwAGwAYQB5AE4AYQBtAGUAIgBdADsACgBpAGYAIAAoACIAZABpAHMAcABsAGEAeQBOAGEAbQBlACIAIABpAG4AIABzAGUAdAB0AGkAbgBnAHMAKQAgAHsACgAgACAAYwBvAG4AcwBvAGwAZQAuAGwAbwBnACgAYABZAG8AdQAgAGgAYQB2AGUAIABhACAAZABpAHMAcABsAGEAeQAgAG4AYQBtAGUALAAgACQAewBzAGUAdAB0AGkAbgBnAHMAWwAiAGQAaQBzAHAAbABhAHkATgBhAG0AZQAiAF0AfQAhAGAAKQA7AAoAfQAgAGUAbABzAGUAIAB7AAoAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoACIAWQBvAHUAIABkAG8AbgAnAHQAIABoAGEAdgBlACAAYQAgAGQAaQBzAHAAbABhAHkAIABuAGEAbQBlACwAIAB3AG8AdQBsAGQAIAB5AG8AdQAgAGwAaQBrAGUAIAB0AG8AIABzAGUAdAAgAG8AbgBlAD8AIgApADsACgB9AAoA)

### Expected result

I would strongly prefer it if Rome did not error or warn on the `delete` operator by default.

As in the example above, it is a common pattern to use an object (Typescript `Record`) to hold several potential fields, and for the *absence* of a field to be treated differently from that value being set to `undefined`. In the example above, `"displayName" in settings` is a fairly common way to test whether a field is present.

If the `delete` line is replaced with `settings["displayName"] = undefined;`, then the wrong `if` branch is taken, which is very undesirable. Even if it's a suggestion rather than an automatic fix, I would be worried that this could introduce bugs. Even a fairly thorough test suite might miss a code replacement bug like this, given that this particularly concerns edge cases for `if` statements.

In theory, `"displayName" in settings` could be replaced with a different check. One "obvious" initial replacement is to do `if(settings.displayName)`, but this would be incorrect in cases where an empty string or another false-ish value (e.g. `0` for a number field) should not be treated the same as a missing field. A "proper" fix would be `typeof settings.displayName !== "undefined"`, but 1) this is significantly less obvious, 2) Rome does not recommend this at the moment, and 3) this can have unintended effects like affecting bundle size if the code contains a lot of such checks.

In addition, adding `undefined` values to fields can lead to other surprising behaviour.

- JSON objects are used throughout various tools in the ecosystem, which may also not treat `undefined` the same as a missing field by default. In particular, other parsers and schemas might *not* accept such values, and the suggested fix could cause breakage.
- Beyond outright breakage, setting a field to `undefined` can unexpectedly change the serialization or key set of an object, which may cause equality checks and optimizations to fail, and audit logs to reflect "ghost" changes.
- If an object contains a lot of fields, this could cause churn and bloat in values. A settings object with a hundred optional fields might end up being serialized into something an order of magnitude larger than intended.

In essence, I am advocating that it is common and reasonable for codebases to stick with the convention that fields should be `delete`d rather than `undefined`, and that the default/recommended Rome rules should not cause friction with this convention. The simplest way to achieve this would be to make `correctness/noDelete` opt-in.

At the moment, I expect to disable this rule for every project, but 1) removing the rule by default would help for [files outside projects with an explicit `rome.json` file](https://github.com/rome/tools/issues/3506), and 2) I would really love to be able to use and recommend Rome defaults without a "gotcha" like this.

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

**Comment by MichaReiser** [comment]
Thanks @lgarron for starting this discussion and for your extensive explanation. You raise many valid points. 

What I take from your explanation is that there are three issues:

a) Unclear fix: The rule doesn't make it clear what the consequences of the suggested fix are and fails to point out other alternatives (using spread `({displayName: _, ...settings} = settings)`) that maintain the same shape as when using `delete` (but has other downsides
b) The fact that the rule is recommended by default. 
c) Unclear motivation

My experience has been that I only rarely needed the `delete` operator and there have oftentimes been alternatives to achieve the same. In my view, the main issue with the delete operation is its [negative impact](https://webkit.org/blog/10298/inline-caching-delete/) on performance-sensitive code. We probably don't do a good enough job explaining this today, and it may be worth moving the rule to a `performance` group instead. 


> JSON objects are used throughout various tools in the ecosystem, which may also not treat undefined the same as a missing field by default. In particular, other parsers and schemas might not accept such values, and the suggested fix could cause breakage.

The default behaviour of JSON is to ignore undefined

```
JSON.stringify({ a: undefined })
> "{}" 
```

What are the tools that you use that deviate from this behavior?

**Reply by lgarron** [reply]
> My experience has been that I only rarely needed the `delete` operator and there have oftentimes been alternatives to achieve the same. In my view, the main issue with the delete operation is its [negative impact](https://webkit.org/blog/10298/inline-caching-delete/) on performance-sensitive code. We probably don't do a good enough job explaining this today, and it may be worth moving the rule to a `performance` group instead.

Yeah, I think this would make a lot more sense.

You make some good points, and I feel like I've learned some things in this very thread! But I'm still rather concerned about correctness within a running JS environment, where `"a" in b` and `typeof a["b"] !== "undefined"` are not the same thing. I think I'd be less concerned if Rome explained the performance concerns and how to fix it without risking correctness issues in other parts of the program.

(FWIW, in the main codebase I'm using with `rome`, we never use `delete` in a hot path.)

**Reply by MichaReiser** [reply]
Sounds good. Let me create an issue on Monday to move the rule to a new performance group because the rule isn't about correctness (it doesn't fix any potential bugs). Part of that will be to better explain the motivation. 

I'll create another issue to suggest destructuring as an alternative fix.

**Comment by ghost** [comment]
Amazing result works.

**Comment by dhirajarun** [comment]
[Playground Link](https://docs.rome.tools/playground/?code=aQBuAHQAZQByAGYAYQBjAGUAIABBAGMAYwBvAHUAbgB0AFMAZQB0AHQAaQBuAGcAcwAgAHsACgAgACAAdQBzAGUAcgBJAEQAOgAgAHMAdAByAGkAbgBnADsACgAgACAAZABpAHMAcABsAGEAeQBOAGEAbQBlAD8AOgAgAHMAdAByAGkAbgBnADsACgB9AAoACgBjAG8AbgBzAHQAIABzAGUAdAB0AGkAbgBnAHMAOgAgAEEAYwBjAG8AdQBuAHQAUwBlAHQAdABpAG4AZwBzACAAPQAgAHsACgAgACAAdQBzAGUAcgBJAEQAOgAgACIAbABnAGEAcgByAG8AbgAiACwACgAgACAAZABpAHMAcABsAGEAeQBOAGEAbQBlADoAIAAiAEwAdQBjAGEAcwAgAEcAYQByAHIAbwBuACIALAAKAH0AOwAKAGQAZQBsAGUAdABlACAAcwBlAHQAdABpAG4AZwBzAFsAIgBkAGkAcwBwAGwAYQB5AE4AYQBtAGUAIgBdADsAIAAvAC8AIACQIZAhkCGQISAAdABoAGkAcwAgAGwAaQBuAGUAIQAKAGkAZgAgACgAIgBkAGkAcwBwAGwAYQB5AE4AYQBtAGUAIgAgAGkAbgAgAHMAZQB0AHQAaQBuAGcAcwApACAAewAKACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKABgAFkAbwB1ACAAaABhAHYAZQAgAGEAIABkAGkAcwBwAGwAYQB5ACAAbgBhAG0AZQAsACAAJAB7AHMAZQB0AHQAaQBuAGcAcwBbACIAZABpAHMAcABsAGEAeQBOAGEAbQBlACIAXQB9ACEAYAApADsACgB9ACAAZQBsAHMAZQAgAHsACgAgACAAYwBvAG4AcwBvAGwAZQAuAGwAbwBnACgAIgBZAG8AdQAgAGQAbwBuACcAdAAgAGgAYQB2AGUAIABhACAAZABpAHMAcABsAGEAeQAgAG4AYQBtAGUALAAgAHcAbwB1AGwAZAAgAHkAbwB1ACAAbABpAGsAZQAgAHQAbwAgAHMAZQB0ACAAbwBuAGUAPwAiACkAOwAKAH0A&jsx=false)

**Comment by ricardobeat** [comment]
@MichaReiser I see the `noDelete` rule was moved under _Performance_ but there is no explanation of why it is needed.

The only instances I have in my codebase are ones where I need `"key" in foo` to return false. This rule causes `--apply-suggested` to become a dangerous transform and not a simple style correction, and also means migration from `prettier` involves more risk than necessary. With this in mind I also support removing it from the defaults.

**Reply by leops** [reply]
This is probably not something we've made clear enough but `--apply-suggested` is indeed intended to be potentially dangerous to apply, if you want to only apply safe fixes you should be using `--apply` instead. I think someone mentioned the idea of renaming `--apply-suggested` to `--apply-unsafe` to make it clearer but unfortunately that's something we haven't done yet, in part because this option isn't necessarily intended to stay (eventually we want to have the CLI prompt the user before applying suggested fixes so they can be manually reviewed)

**Comment by malobre** [comment]
While this might be important to consider in a hot-loop, it can be totally fine in other cases (e.g: removing an item from a cart in a click event handler). Maybe this lint level could be lowered to a warning ?

**Comment by Spookerton** [comment]
noDelete was added a few years back in PR <https://github.com/rome/tools/pull/350>.
The reasoning is listed in doc as "discussed in discord". I searched based on the PR date and found <https://discord.com/channels/678763474494423051/678763474930761739/704479585877753967>. I don't intend to dig further. The example given is of using delete on array entries, resulting in those entries being set to undefined.

The delete keyword is intended for use with object properties. Its behavior on array entries is only understandable - not exactly intended.

Here's a very MRE:
```js
const cache = Object.create(null)
function cachePut (key, value) { cache[key] = value }
function cacheGet (key) { return cache[key] }
function cacheDel (key) { delete cache[key] }
```

And something to paste into a .mjs:
```js
class Cache {
  #lifespan = undefined
  #entries = Object.create(null)
  constructor (lifespan) {
    this.#lifespan = lifespan
  }
  get (key) {
    return this.#entries[key]?.value
  }
  put (key, value) {
    this.#entries[key] = Object.assign(Object.create(null), {
      expires: Date.now() + this.#lifespan,
      value
    })
  }
  del (key) {
    delete this.#entries[key]
  }
  clean (hard) {
    if (hard) {
      this.#entries = Object.create(null)
      return
    }
    const time = Date.now()
    for (const key in this.#entries)
      if (this.#entries[key].expires < time)
        delete this.#entries[key]
  }
}

const blah = new Cache(1000)
blah.put("cat", "dog")
console.log(blah.get("cat"))
blah.clean()
console.log(blah.get("cat"))
await new Promise((done) => setTimeout(done, 2000))
blah.clean()
console.log(blah.get("cat"))
```

Especially in long life processes, node, deno, or bun, etc, this kind of object-backed cache is common for holding ephemeral data, such as (source => token) pairs for logins.

The reason the noDelete rule is a problem in this circumstance is that assigning undefined to a key *does not delete the key*. Over the lifespan of a long-lived app, such an object may easily accumulate a large hidden debt of keys -- and stringification in common JS engines actually hides this problem by being clever.

As a very simple example with noise removed:

```
λ node
> const foo = {}
> foo.bar = { name: 'spot' }
> delete foo.bar
> JSON.stringify(foo)
'{}'
> for (const key in foo) console.log(key)
>

λ node
> const foo = {}
> foo.bar = { name: 'spot' }
> foo.bar = undefined
> JSON.stringify(foo)
'{}'
> for (const key in foo) console.log(key)
bar
>
```

The application language use case conflicts with the article mentioned in <https://github.com/rome/tools/discussions/3614#discussioncomment-4117626>. The reason for that is that the two targets are very different - The article talks about the delete keyword negatively impacting webgl rendering code with crafted examples. I can see that being a thing. I don't know where the tradeoff begins.

One might circumvent the warning in either case by using Map or recomposition and replacement.

I believe the rule is too broad and not valid, except in reference to the original part of this post, for style reasons - no delete on array indices.

**Comment by Conaclos** [comment]
Taking feedbacks into account I could suggest relaxing the rule: we could allow `delete` on `JsComputedMemberExpression` with no literal keys.

The following pattern would pass:

```js
delete record[computed]
```

While the following patterns would still be rejected:

```js
delete record["literal"]
delete record[0]
delete record.a
```

**Reply by lgarron** [reply]
Unfortunately, rejecting `delete record["literal"]` or `delete record.a
 ` would still cause the issues described in my original post. 😕

**Reply by Conaclos** [reply]
@lgarron Have you some real code example of this kind of usage?

**Reply by lgarron** [reply]
https://github.com/cubing/cubing.js/ currently uses this pattern in 8 places.

**Reply by Conaclos** [reply]
You can avoid all `delete obj.prop` by relying on destructring.

In `src/cubing/search/outside.ts`

```diff
- const { ...def } = kpuzzle.definition;
- delete def.experimentalIsStateSolved;
+ const { experimentalIsStateSolved, ...def } = kpuzzle.definition;
```

This avoids mutating the object shape, and is shorter.

In the same vein in `src/cubing/vendor/mpl/twsearch/twsearch-BDAXZGZU.js`:

```diff
- const config = getConfigFromURL("", this.a.href);
+ const { puzzle, ...config} = getConfigFromURL("", this.a.href);

- if (config.puzzle && !(config.puzzle in puzzles)) {
+ if (puzzle && !(puzzle in puzzles)) {
```

This is also shorter.

This has also the advantage to avoid a mutation that is a common assumption for config objects...

**Reply by lgarron** [reply]
> You can avoid all `delete obj.prop` by relying on destructring.

This is not true in general, if you are modifying an object that you do not have the only reference to.

> This avoids mutating the object shape, and is shorter.

I'll grant you that it works here, but 1) sometimes mutation *is* the point, and 2) while I'm a big fan of destructuring, I'm not sure I the result is easier to understand, especially for people who are not JS nerds.

Also, how would you use this to delete an entry from `localStorage`?

**Comment by malobre** [comment]
Another case where you can't avoid `delete` is removing a data attribute from an HTML element. The `dataset` property is read-only and values are automatically serialized (even `null` and `undefined`), using `delete` is the only way to remove a custom attribute.

Edit: I'm wrong, there is one other way: you can use `element.removeAttribute("data-my-attribute")`.
